components/org.apache.stratos.rest.endpoint/src/main/java/org/apache/stratos/rest/endpoint/api/StratosApiV41Utils.java

```diff
@@ -980,50 +981,41 @@ public static void addServiceGroup(GroupBean serviceGroupDefinition) throws Rest
             List<String> groupNames;
             String[] cartridgeGroupNames;
 
-            // if any cartridges are specified in the group, they should be already deployed
-            if (serviceGroupDefinition.getCartridges() != null) {
+            if (log.isDebugEnabled()) {
+                log.debug("checking cartridges in cartridge group " + serviceGroupDefinition.getName());
+            }
 
-                if (log.isDebugEnabled()) {
-                    log.debug("checking cartridges in cartridge group " + serviceGroupDefinition.getName());
-                }
+            findCartridgesInGroupBean(serviceGroupDefinition, cartridgeTypes);
 
-                findCartridgesInGroupBean(serviceGroupDefinition, cartridgeTypes);
+            //validate the group definition to check if cartridges duplicate in any groups defined
+            validateCartridgeDuplicationInGroupDefinition(serviceGroupDefinition);
 
-                Set<String> duplicates = findDuplicates(cartridgeTypes);
-                if (duplicates.size() > 0) {
-                    StringBuilder duplicatesOutput = new StringBuilder();
-                    for (String dup : duplicates) {
-                        duplicatesOutput.append(dup).append(" ");
-                    }
-                    if (log.isDebugEnabled()) {
-                        log.debug("duplicate cartridges defined: " + duplicatesOutput.toString());
-                    }
-                    throw new RestAPIException("Invalid Service Group definition, duplicate cartridges defined:" +
-                            duplicatesOutput.toString());
-                }
+            //validate the group definition to check if groups duplicate in any groups and
+            //validate the group definition to check for cyclic group behaviour
+            validateGroupDuplicationInGroupDefinition(serviceGroupDefinition, new ArrayList<String>());
```

Comment:
Still I think we could introduce a method overload to avoid passing a new object of ArrayList<String>:
validateGroupDuplicationInGroupDefinition(serviceGroupDefinition)